jc/parsers/ascii.py

```diff
@@ -0,0 +1,178 @@
+"""jc - JSON CLI output utility ASCII table parser
+
+Supports common ASCII table formats
+
+Usage (cli):
+
+    $ virsh list --all | jc --ascii
+
+    or
+
+    $ jc ascii
+
+Usage (module):
+
+    import jc.parsers.ascii
+    result = jc.parsers.ascii.parse(ascii_table_output)
+
+Schema:
+
+    [
+      {
+        "column1":     string,
+        "column2":     boolean,
+        "column3":     integer
+      }
+    ]
+    
+Examples:
+
+    $ virsh domifaddr foo | jc --ascii
+    [{"name":"vnet0","mac_address":"01:23:45:67:89:ab","protocol":"ipv4","address":"192.168.122.100/24"}]
+
+"""
+import jc.utils
+import jc.parsers.universal
+import re
+
+
+class info():
+    """Provides parser metadata (version, author, etc.)"""
+    version = '1.0'
+    description = 'ASCII table parser'
+    author = 'Alexandru Barbur'
+    author_email = 'alex@ctrlc.name'
+    # details = 'enter any other details here'
+
+    # compatible options: linux, darwin, cygwin, win32, aix, freebsd
+    compatible = ['linux', 'darwin', 'cygwin', 'win32', 'aix', 'freebsd']
+    magic_commands = ['ascii']
+
+
+__version__ = info.version
+
+
+def _process(proc_data):
+    """
+    Final processing to conform to the schema.
+
+    Parameters:
+
+        proc_data:   (List of Dictionaries) raw structured data to process
+
+    Returns:
+
+        List of Dictionaries. Structured to conform to the schema.
+    """
+
+    # rebuild output for added semantic information
+    # use helper functions in jc.utils for int, float, bool conversions and timestamps
+
+    return proc_data
+
+
+def _snake_case(value):
+    """
+    Convert a value to snake case
+    
+    Parameters:
+    
+        value:    (string) original value
+        
+    Returns:
+    
+        Snake cased value.
+    """
+
+    value = value.replace('-', '_')
+    value = re.sub(r'([A-Z]+)([A-Z][a-z])', r'\1_\2', value)
+    value = re.sub(r'([a-z\d])([A-Z])', r'\1_\2', value)
+    value = re.sub(r'\s+', r'_', value)
+    return value.lower()
+
+
+def _process_headers(line):
+    """
+    Detect and clean header values in the table header row
+    
+    Parameters:
+    
+        line:    (string) header row line data
+        
+    Returns:
+    
+        Line with cleaned header values.
+    """
+
+    # XXX: what about tables that only use a single space between columns?
+    matches = list(re.finditer(r"\w+( \w+)*", line))
```

Comment:
This regular expression is used to detect header values.

Reply:
I'll take a closer look at the regex itself, but any reason not to use `re.findall()` instead of `re.finditer()` since it is being cast to a list anyway?

Reply:
IIRC it's because `findall()` returns tuples without the start and end positions while `finditer()` returns match objects that do include those. Compare the output:

```
>>> re.findall(r'\w+( \w+)*', ' Hello World    This is cool    Test')
[' World', ' cool', '']
>>> list(re.finditer(r'\w+( \w+)*', ' Hello World    This is cool    Test'))
[<re.Match object; span=(1, 12), match='Hello World'>, <re.Match object; span=(16, 28), match='This is cool'>, <re.Match object; span=(32, 36), match='Test'>]
>>>
```

Reply:
I could probably take out the `list()` call though and iterate using the iterator returned directly. Think that was left over from debugging.

Reply:
Interesting  - I'll have to wrap my head around why the output is different with the same regex pattern, but I see you are using the start/end attributes that come with `re.finditer` so that makes sense.